File: Creational/AbstractFactory/CSharp/TwoDShapeFactory.cs
// TwoDShapeFactory.cs
using System;

public class TwoDShapeFactory : IShapeFactory {
  public IShape CreateShape(ShapeType type) {
    return type switch {
      ShapeType.CIRCLE => new Circle(),
      ShapeType.SQUARE => new Square(),
      _ => throw new ArgumentException($"Unknown shape type: {type}")
    };
  }

  public IRenderer CreateRenderer() {
    return new TwoDRenderer();
  }
}

File: Creational/AbstractFactory/Cpp/TwoDShapeFactory.cpp
// TwoDShapeFactory.cpp

#include "TwoDShapeFactory.h"
#include "Circle.h"
#include "Square.h"
#include "TwoDRenderer.h"
#include <stdexcept>

IShape* TwoDShapeFactory::createShape(ShapeType type) {
  switch (type) {
    case ShapeType::CIRCLE:
      return new Circle();
    case ShapeType::SQUARE:
      return new Square();
    default:
      throw std::invalid_argument("Unknown shape type");
  }
}

IRenderer* TwoDShapeFactory::createRenderer() {
  return new TwoDRenderer();
}
